rllib/utils/replay_buffers/storage.py

```diff
@@ -0,0 +1,503 @@
+        self._num_timesteps += item.count
+        self._size_bytes += item.size_bytes()
+
+        # Drop old items.
+        # May require multiple drops if newly added item
+        # contains more timesteps than the old items.
+        while self._num_timesteps > self.capacity:
+            assert self._num_items > 0
+            self._eviction_started = True
+            self._evicted_hit_stats.push(self._hit_count[self._offset_idx])
+            self._hit_count[self._offset_idx] = 0
+            drop_item = self._del(self._offset_idx)
+            self._num_timesteps -= drop_item.count
+            self._size_bytes -= drop_item.size_bytes()
+            self._num_items -= 1
+            self._offset_idx = self._get_internal_index(1)  # Increase offset
+
+        # Insert new item.
+        # Compute index to set new item at in circular storage.
+        # Wrap around once we hit capacity.
+        new_idx = self._get_internal_index(self._num_items)
+        self._set(new_idx, item)
+        self._num_items += 1
+
+    def _get_internal_index(self, idx: int):
+        """Translate the given external storage index into
+        the internal index space of the circular buffer.
+
+        Args:
+            idx: External storage index (0 <= idx < len(storage)).
+
+        Returns:
+            Internal index from interval [0, capacity)
+        """
+        if idx < 0:
+            raise IndexError('Buffer index out of range')
+        return (self._offset_idx + idx) % self.capacity
+
+    def _get_external_index(self, idx: int):
+        """Translate the given internal circular buffer index into
+        the external index space of the storage.
+
+        Args:
+            idx: Internal circular Buffer index (0 <= idx < capacity).
+
+        Returns:
+            External index from interval [0, len(storage))
+        """
+        if idx < 0:
+            raise IndexError('Buffer index out of range')
+        if idx >= self._offset_idx:
+            return idx - self._offset_idx
+        else:
+            return idx + self.capacity - self._offset_idx
+
+    @abstractmethod
+    def _get(self, idx: int) -> SampleBatchType:
+        """Get the item at the specified index / key.
+
+        This method must be implementend by subclasses
+        using an actual data structure for storing the data.
+        This data structure must be capable of dealing with
+        indices between 0 <= idx < capacity.
+
+        Args:
+            idx: Index of the item of interest.
+
+        Returns:
+            Item at index.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def _set(self, idx: int, item: SampleBatchType) -> None:
+        """Store the given item at the specified index / key.
+
+        This method must be implementend by subclasses
+        using an actual data structure for storing the data.
+        This data structure must be capable of dealing with
+        indices between 0 <= idx < capacity.
```

Comment:
What about items that are stored as sequences or episodes? The constructor says timesteps only and that's also how I read the add() method.

Reply:
Storage is just a replacement for the list currently used in ReplayBuffer. Whatever is stored in the list currently, will now be stored in Storage. The structuring / splitting the items is up to the ReplayBuffer.